In [30]:
import pandas as pd
import numpy as np
from matplotlib import pyplot as plt
import seaborn as sns
sns.set()

In [31]:
application_details = pd.read_csv('./data/application_record 2.csv')
credit_records = pd.read_csv('./data/credit_record.csv')

In [32]:
application_details.describe().T

count          mean            std        min        25%  \
ID                438557.0  6.022176e+06  571637.023257  5008804.0  5609375.0   
CNT_CHILDREN      438557.0  4.273903e-01       0.724882        0.0        0.0   
AMT_INCOME_TOTAL  438557.0  1.875243e+05  110086.853066    26100.0   121500.0   
DAYS_BIRTH        438557.0 -1.599790e+04    4185.030007   -25201.0   -19483.0   
DAYS_EMPLOYED     438557.0  6.056368e+04  138767.799647   -17531.0    -3103.0   
FLAG_MOBIL        438557.0  1.000000e+00       0.000000        1.0        1.0   
FLAG_WORK_PHONE   438557.0  2.061328e-01       0.404527        0.0        0.0   
FLAG_PHONE        438557.0  2.877710e-01       0.452724        0.0        0.0   
FLAG_EMAIL        438557.0  1.082071e-01       0.310642        0.0        0.0   
CNT_FAM_MEMBERS   438557.0  2.194465e+00       0.897207        1.0        2.0   

                        50%        75%        max  
ID                6047745.0  6456971.0  7999952.0  
CNT_CHILDREN            0.0        1.0       19.0  
AMT_INCOME_TOTAL   160780.5   225000.0  6750000.0  
DAYS_BIRTH         -15630.0   -12514.0    -7489.0  
DAYS_EMPLOYED       -1467.0     -371.0   365243.0  
FLAG_MOBIL              1.0        1.0        1.0  
FLAG_WORK_PHONE         0.0        0.0        1.0  
FLAG_PHONE              0.0        1.0        1.0  
FLAG_EMAIL              0.0        0.0        1.0  
CNT_FAM_MEMBERS         2.0        3.0       20.0

# Inspecting Data

In [33]:
credit_records.describe().T

count          mean           std        min        25%  \
ID              1048575.0  5.068286e+06  46150.578505  5001711.0  5023644.0   
MONTHS_BALANCE  1048575.0 -1.913700e+01     14.023498      -60.0      -29.0   

                      50%        75%        max  
ID              5062104.0  5113856.0  5150487.0  
MONTHS_BALANCE      -17.0       -7.0        0.0

In [34]:
application_details.tail()

ID CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  CNT_CHILDREN  \
438552  6840104           M            N               Y             0   
438553  6840222           F            N               N             0   
438554  6841878           F            N               N             0   
438555  6842765           F            N               Y             0   
438556  6842885           F            N               Y             0   

        AMT_INCOME_TOTAL      NAME_INCOME_TYPE            NAME_EDUCATION_TYPE  \
438552          135000.0             Pensioner  Secondary / secondary special   
438553          103500.0               Working  Secondary / secondary special   
438554           54000.0  Commercial associate               Higher education   
438555           72000.0             Pensioner  Secondary / secondary special   
438556          121500.0               Working  Secondary / secondary special   

          NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  DAYS_BIRTH  DAYS_EMPLOYED  \
438552             Separated  House / apartment      -22717         365243   
438553  Single / not married  House / apartment      -15939          -3007   
438554  Single / not married       With parents       -8169           -372   
438555               Married  House / apartment      -21673         365243   
438556               Married  House / apartment      -18858          -1201   

        FLAG_MOBIL  FLAG_WORK_PHONE  FLAG_PHONE  FLAG_EMAIL OCCUPATION_TYPE  \
438552           1                0           0           0             NaN   
438553           1                0           0           0        Laborers   
438554           1                1           0           0     Sales staff   
438555           1                0           0           0             NaN   
438556           1                0           1           0     Sales staff   

        CNT_FAM_MEMBERS  
438552              1.0  
438553              1.0  
438554              1.0  
438555              2.0  
438556              2.0

In [35]:
credit_records.tail()

ID  MONTHS_BALANCE STATUS
1048570  5150487             -25      C
1048571  5150487             -26      C
1048572  5150487             -27      C
1048573  5150487             -28      C
1048574  5150487             -29      C

In [36]:
np.unique(credit_records['STATUS'])

array(['0', '1', '2', '3', '4', '5', 'C', 'X'], dtype=object)

- According to source data, credit_records has the following information:
Feature name	Explanation	Remarks
ID	Client number	
- 0 is the current month, 
- -1 is the previous month, and so on
STATUS	Status	
- 0: 1-29 days past due 
- 1: 30-59 days past due 
- 2: 60-89 days overdue 
- 3: 90-119 days overdue 
- 4: 120-149 days overdue 
- 5: Overdue or bad debts, 
- write-offs for more than 150 days 
- C: paid off that month 
- X: No loan for the month

# Data Wrangling

 **Converting credit_record status to binary(1/0) data**


In [37]:
convert_to = {'C' : 'Good_Debt', 'X' : 'Good_Debt', '0' : 'Good_Debt', '1' : 'Neutral_Debt', '2' : 'Neutral_Debt', '3' : 'Bad_Debt', '4' : 'Bad_Debt', '5' : 'Bad_Debt'}
credit_records.replace({'STATUS' : convert_to}, inplace=True)

In [38]:
credit_records

ID  MONTHS_BALANCE     STATUS
0        5001711               0  Good_Debt
1        5001711              -1  Good_Debt
2        5001711              -2  Good_Debt
3        5001711              -3  Good_Debt
4        5001712               0  Good_Debt
...          ...             ...        ...
1048570  5150487             -25  Good_Debt
1048571  5150487             -26  Good_Debt
1048572  5150487             -27  Good_Debt
1048573  5150487             -28  Good_Debt
1048574  5150487             -29  Good_Debt

[1048575 rows x 3 columns]

- Convert status into approval status

In [39]:
credit_records.loc[credit_records['STATUS'] == 'Good_Debt', 'APPROVAL_STATUS'] = 1
credit_records.loc[credit_records['STATUS'] == 'Bad_Debt', 'APPROVAL_STATUS'] = 0


In [11]:
credit_records

ID  MONTHS_BALANCE     STATUS  APPROVAL_STATUS
0        5001711               0  Good_Debt              1.0
1        5001711              -1  Good_Debt              1.0
2        5001711              -2  Good_Debt              1.0
3        5001711              -3  Good_Debt              1.0
4        5001712               0  Good_Debt              1.0
...          ...             ...        ...              ...
1048570  5150487             -25  Good_Debt              1.0
1048571  5150487             -26  Good_Debt              1.0
1048572  5150487             -27  Good_Debt              1.0
1048573  5150487             -28  Good_Debt              1.0
1048574  5150487             -29  Good_Debt              1.0

[1048575 rows x 4 columns]

In [40]:
application_details.isnull().sum()

ID                          0
CODE_GENDER                 0
FLAG_OWN_CAR                0
FLAG_OWN_REALTY             0
CNT_CHILDREN                0
AMT_INCOME_TOTAL            0
NAME_INCOME_TYPE            0
NAME_EDUCATION_TYPE         0
NAME_FAMILY_STATUS          0
NAME_HOUSING_TYPE           0
DAYS_BIRTH                  0
DAYS_EMPLOYED               0
FLAG_MOBIL                  0
FLAG_WORK_PHONE             0
FLAG_PHONE                  0
FLAG_EMAIL                  0
OCCUPATION_TYPE        134203
CNT_FAM_MEMBERS             0
dtype: int64

- about one third of the record has null value in OCCUPATION_TYPE column

In [41]:
application_details = application_details.dropna()

In [42]:
application_details.isnull().sum()

ID                     0
CODE_GENDER            0
FLAG_OWN_CAR           0
FLAG_OWN_REALTY        0
CNT_CHILDREN           0
AMT_INCOME_TOTAL       0
NAME_INCOME_TYPE       0
NAME_EDUCATION_TYPE    0
NAME_FAMILY_STATUS     0
NAME_HOUSING_TYPE      0
DAYS_BIRTH             0
DAYS_EMPLOYED          0
FLAG_MOBIL             0
FLAG_WORK_PHONE        0
FLAG_PHONE             0
FLAG_EMAIL             0
OCCUPATION_TYPE        0
CNT_FAM_MEMBERS        0
dtype: int64

In [43]:
application_details

ID CODE_GENDER FLAG_OWN_CAR FLAG_OWN_REALTY  CNT_CHILDREN  \
2       5008806           M            Y               Y             0   
3       5008808           F            N               Y             0   
4       5008809           F            N               Y             0   
5       5008810           F            N               Y             0   
6       5008811           F            N               Y             0   
...         ...         ...          ...             ...           ...   
438541  6837707           M            N               Y             0   
438548  6839936           M            Y               Y             1   
438553  6840222           F            N               N             0   
438554  6841878           F            N               N             0   
438556  6842885           F            N               Y             0   

        AMT_INCOME_TOTAL      NAME_INCOME_TYPE            NAME_EDUCATION_TYPE  \
2               112500.0               Working  Secondary / secondary special   
3               270000.0  Commercial associate  Secondary / secondary special   
4               270000.0  Commercial associate  Secondary / secondary special   
5               270000.0  Commercial associate  Secondary / secondary special   
6               270000.0  Commercial associate  Secondary / secondary special   
...                  ...                   ...                            ...   
438541          202500.0               Working               Higher education   
438548          135000.0               Working  Secondary / secondary special   
438553          103500.0               Working  Secondary / secondary special   
438554           54000.0  Commercial associate               Higher education   
438556          121500.0               Working  Secondary / secondary special   

          NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  DAYS_BIRTH  DAYS_EMPLOYED  \
2                    Married  House / apartment      -21474          -1134   
3       Single / not married  House / apartment      -19110          -3051   
4       Single / not married  House / apartment      -19110          -3051   
5       Single / not married  House / apartment      -19110          -3051   
6       Single / not married  House / apartment      -19110          -3051   
...                      ...                ...         ...            ...   
438541        Civil marriage  House / apartment      -13510          -2309   
438548               Married  House / apartment      -12569          -2095   
438553  Single / not married  House / apartment      -15939          -3007   
438554  Single / not married       With parents       -8169           -372   
438556               Married  House / apartment      -18858          -1201   

        FLAG_MOBIL  FLAG_WORK_PHONE  FLAG_PHONE  FLAG_EMAIL OCCUPATION_TYPE  \
2                1                0           0           0  Security staff   
3                1                0           1           1     Sales staff   
4                1                0           1           1     Sales staff   
5                1                0           1           1     Sales staff   
6                1                0           1           1     Sales staff   
...            ...              ...         ...         ...             ...   
438541           1                1           0           0        Laborers   
438548           1                0           0           0        Laborers   
438553           1                0           0           0        Laborers   
438554           1                1           0           0     Sales staff   
438556           1                0           1           0     Sales staff   

        CNT_FAM_MEMBERS  
2                   2.0  
3                   1.0  
4                   1.0  
5                   1.0  
6                   1.0  
...                 ...  
438541              2.0  
438548              3.0  
438553              1.0  
43855

**Convert days data to readable years format**

In [44]:
application_details['AGE'] = np.ceil(pd.to_timedelta(application_details['DAYS_BIRTH'], unit='D').dt.days / -365.25)

In [45]:
application_details.drop('DAYS_BIRTH', axis=1, inplace=True)

In [46]:
# Checking Data after converting DAYS_BIRTH to AGES
application_details.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 304354 entries, 2 to 438556
Data columns (total 18 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   ID                   304354 non-null  int64  
 1   CODE_GENDER          304354 non-null  object 
 2   FLAG_OWN_CAR         304354 non-null  object 
 3   FLAG_OWN_REALTY      304354 non-null  object 
 4   CNT_CHILDREN         304354 non-null  int64  
 5   AMT_INCOME_TOTAL     304354 non-null  float64
 6   NAME_INCOME_TYPE     304354 non-null  object 
 7   NAME_EDUCATION_TYPE  304354 non-null  object 
 8   NAME_FAMILY_STATUS   304354 non-null  object 
 9   NAME_HOUSING_TYPE    304354 non-null  object 
 10  DAYS_EMPLOYED        304354 non-null  int64  
 11  FLAG_MOBIL           304354 non-null  int64  
 12  FLAG_WORK_PHONE      304354 non-null  int64  
 13  FLAG_PHONE           304354 non-null  int64  
 14  FLAG_EMAIL           304354 non-null  int64  
 15  OCCUPATION_TYPE  

In [47]:
application_details.loc[(application_details['DAYS_EMPLOYED'] >= 0), 'DAYS_EMPLOYED'] = 0
application_details['YEARS_EMPLOYED'] = np.ceil(pd.to_timedelta(application_details['DAYS_EMPLOYED'], unit='D').dt.days / -365.25)


In [48]:
application_details.drop('DAYS_EMPLOYED', axis=1, inplace=True)


In [49]:
# Checking Data after converting DAYS_BIRTH to AGES
application_details.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 304354 entries, 2 to 438556
Data columns (total 18 columns):
 #   Column               Non-Null Count   Dtype  
---  ------               --------------   -----  
 0   ID                   304354 non-null  int64  
 1   CODE_GENDER          304354 non-null  object 
 2   FLAG_OWN_CAR         304354 non-null  object 
 3   FLAG_OWN_REALTY      304354 non-null  object 
 4   CNT_CHILDREN         304354 non-null  int64  
 5   AMT_INCOME_TOTAL     304354 non-null  float64
 6   NAME_INCOME_TYPE     304354 non-null  object 
 7   NAME_EDUCATION_TYPE  304354 non-null  object 
 8   NAME_FAMILY_STATUS   304354 non-null  object 
 9   NAME_HOUSING_TYPE    304354 non-null  object 
 10  FLAG_MOBIL           304354 non-null  int64  
 11  FLAG_WORK_PHONE      304354 non-null  int64  
 12  FLAG_PHONE           304354 non-null  int64  
 13  FLAG_EMAIL           304354 non-null  int64  
 14  OCCUPATION_TYPE      304354 non-null  object 
 15  CNT_FAM_MEMBERS  

In [50]:
def Cat_to_Num(features):
    for feature in features:
        feature_list = list(np.unique(application_details[feature]))
        feature_dict = {}
        for i in range(len(feature_list)):
                       feature_dict[feature_list[i]] = i
        application_details.replace({feature : feature_dict}, inplace=True)
        print(feature, '-->', feature_dict)

In [51]:
categorical_features = ['CODE_GENDER', 'FLAG_OWN_CAR', 'FLAG_OWN_REALTY', 'NAME_INCOME_TYPE', 'NAME_FAMILY_STATUS', 'OCCUPATION_TYPE', 'NAME_HOUSING_TYPE', 'NAME_EDUCATION_TYPE']
Cat_to_Num(categorical_features)

CODE_GENDER --> {'F': 0, 'M': 1}
FLAG_OWN_CAR --> {'N': 0, 'Y': 1}
FLAG_OWN_REALTY --> {'N': 0, 'Y': 1}
NAME_INCOME_TYPE --> {'Commercial associate': 0, 'Pensioner': 1, 'State servant': 2, 'Student': 3, 'Working': 4}
NAME_FAMILY_STATUS --> {'Civil marriage': 0, 'Married': 1, 'Separated': 2, 'Single / not married': 3, 'Widow': 4}
OCCUPATION_TYPE --> {'Accountants': 0, 'Cleaning staff': 1, 'Cooking staff': 2, 'Core staff': 3, 'Drivers': 4, 'HR staff': 5, 'High skill tech staff': 6, 'IT staff': 7, 'Laborers': 8, 'Low-skill Laborers': 9, 'Managers': 10, 'Medicine staff': 11, 'Private service staff': 12, 'Realty agents': 13, 'Sales staff': 14, 'Secretaries': 15, 'Security staff': 16, 'Waiters/barmen staff': 17}
NAME_HOUSING_TYPE --> {'Co-op apartment': 0, 'House / apartment': 1, 'Municipal apartment': 2, 'Office apartment': 3, 'Rented apartment': 4, 'With parents': 5}
NAME_EDUCATION_TYPE --> {'Academic degree': 0, 'Higher education': 1, 'Incomplete higher': 2, 'Lower secondary': 3, 'Seconda

In [53]:
application_details.head(20)

ID  CODE_GENDER  FLAG_OWN_CAR  FLAG_OWN_REALTY  CNT_CHILDREN  \
2   5008806            1             1                1             0   
3   5008808            0             0                1             0   
4   5008809            0             0                1             0   
5   5008810            0             0                1             0   
6   5008811            0             0                1             0   
10  5008815            1             1                1             0   
11  5112956            1             1                1             0   
12  6153651            1             1                1             0   
13  5008819            1             1                1             0   
14  5008820            1             1                1             0   
15  5008821            1             1                1             0   
16  5008822            1             1                1             0   
17  5008823            1             1                1             0   
18  5008824            1             1                1             0   
19  5008825            0             1                0             0   
20  5008826            0             1                0             0   
21  5008830            0             0                1             0   
22  5008831            0             0                1             0   
23  5008832            0             0                1             0   
27  5008836            1             1                1             3   

    AMT_INCOME_TOTAL  NAME_INCOME_TYPE  NAME_EDUCATION_TYPE  \
2           112500.0                 4                    4   
3           270000.0                 0                    4   
4           270000.0                 0                    4   
5           270000.0                 0                    4   
6           270000.0                 0                    4   
10          270000.0                 4                    1   
11          270000.0                 4                    1   
12          270000.0                 4                    1   
13          135000.0                 0                    4   
14          135000.0                 0                    4   
15          135000.0                 0                    4   
16          135000.0                 0                    4   
17          135000.0                 0                    4   
18          135000.0                 0                    4   
19          130500.0                 4                    2   
20          130500.0                 4                    2   
21          157500.0                 4                    4   
22          157500.0                 4                    4   
23          157500.0                 4                    4   
27          270000.0                 4                    4   

    NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  FLAG_MOBIL  FLAG_WORK_PHONE  \
2                    1                  1           1                0   
3                    3                  1           1                0   
4                    3                  1           1                0   
5                    3                  1           1                0   
6                    3                  1           1                0   
10                   1                  1           1                1   
11                   1                  1           1                1   
12                   1                  1           1                1   
13                   1                  1           1                0   
14                   1                  1           1                0   
15                   1                  1           1                0   
16                   1                  1           1                0   
17                   1                  1           1                0   
18                   1                  1           1                0   
19                   1                  1

In [54]:
credit_data_merged = application_details.merge(credit_records, how='inner', on=['ID'])

In [55]:
credit_data_merged.describe().T

count          mean            std        min  \
ID                   537667.0  5.079231e+06   42001.999788  5008806.0   
CODE_GENDER          537667.0  3.791101e-01       0.485166        0.0   
FLAG_OWN_CAR         537667.0  4.304895e-01       0.495145        0.0   
FLAG_OWN_REALTY      537667.0  6.425371e-01       0.479253        0.0   
CNT_CHILDREN         537667.0  5.066965e-01       0.787285        0.0   
AMT_INCOME_TOTAL     537667.0  1.971171e+05  104138.963465    27000.0   
NAME_INCOME_TYPE     537667.0  2.674957e+00       1.774396        0.0   
NAME_EDUCATION_TYPE  537667.0  3.053723e+00       1.361281        0.0   
NAME_FAMILY_STATUS   537667.0  1.290008e+00       0.867066        0.0   
NAME_HOUSING_TYPE    537667.0  1.308801e+00       0.996071        0.0   
FLAG_MOBIL           537667.0  1.000000e+00       0.000000        1.0   
FLAG_WORK_PHONE      537667.0  2.816148e-01       0.449787        0.0   
FLAG_PHONE           537667.0  2.988932e-01       0.457773        0.0   
FLAG_EMAIL           537667.0  1.007296e-01       0.300971        0.0   
OCCUPATION_TYPE      537667.0  7.617789e+00       4.286959        0.0   
CNT_FAM_MEMBERS      537667.0  2.303069e+00       0.936852        1.0   
AGE                  537667.0  4.159059e+01       9.359462       21.0   
YEARS_EMPLOYED       537667.0  8.065479e+00       6.562763        1.0   
MONTHS_BALANCE       537667.0 -1.930524e+01      14.037827      -60.0   
APPROVAL_STATUS      530702.0  9.973243e-01       0.051658        0.0   

                           25%        50%        75%        max  
ID                   5044925.0  5079091.0  5115755.0  5150487.0  
CODE_GENDER                0.0        0.0        1.0        1.0  
FLAG_OWN_CAR               0.0        0.0        1.0        1.0  
FLAG_OWN_REALTY            0.0        1.0        1.0        1.0  
CNT_CHILDREN               0.0        0.0        1.0       19.0  
AMT_INCOME_TOTAL      135000.0   180000.0   229500.0  1575000.0  
NAME_INCOME_TYPE           0.0        4.0        4.0        4.0  
NAME_EDUCATION_TYPE        1.0        4.0        4.0        4.0  
NAME_FAMILY_STATUS         1.0        1.0        1.0        4.0  
NAME_HOUSING_TYPE          1.0        1.0        1.0        5.0  
FLAG_MOBIL                 1.0        1.0        1.0        1.0  
FLAG_WORK_PHONE            0.0        0.0        1.0        1.0  
FLAG_PHONE                 0.0        0.0        1.0        1.0  
FLAG_EMAIL                 0.0        0.0        0.0        1.0  
OCCUPATION_TYPE            4.0        8.0       10.0       17.0  
CNT_FAM_MEMBERS            2.0        2.0        3.0       20.0  
AGE                       34.0       41.0       49.0       68.0  
YEARS_EMPLOYED             3.0        6.0       11.0       44.0  
MONTHS_BALANCE           -29.0      -17.0       -8.0        0.0  
APPROVAL_STATUS            1.0        1.0        1.0        1.0

In [56]:
credit_data_merged.head()

ID  CODE_GENDER  FLAG_OWN_CAR  FLAG_OWN_REALTY  CNT_CHILDREN  \
0  5008806            1             1                1             0   
1  5008806            1             1                1             0   
2  5008806            1             1                1             0   
3  5008806            1             1                1             0   
4  5008806            1             1                1             0   

   AMT_INCOME_TOTAL  NAME_INCOME_TYPE  NAME_EDUCATION_TYPE  \
0          112500.0                 4                    4   
1          112500.0                 4                    4   
2          112500.0                 4                    4   
3          112500.0                 4                    4   
4          112500.0                 4                    4   

   NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  ...  FLAG_WORK_PHONE  FLAG_PHONE  \
0                   1                  1  ...                0           0   
1                   1                  1  ...                0           0   
2                   1                  1  ...                0           0   
3                   1                  1  ...                0           0   
4                   1                  1  ...                0           0   

   FLAG_EMAIL  OCCUPATION_TYPE  CNT_FAM_MEMBERS   AGE  YEARS_EMPLOYED  \
0           0               16              2.0  59.0             4.0   
1           0               16              2.0  59.0             4.0   
2           0               16              2.0  59.0             4.0   
3           0               16              2.0  59.0             4.0   
4           0               16              2.0  59.0             4.0   

   MONTHS_BALANCE     STATUS APPROVAL_STATUS  
0               0  Good_Debt             1.0  
1              -1  Good_Debt             1.0  
2              -2  Good_Debt             1.0  
3              -3  Good_Debt             1.0  
4              -4  Good_Debt             1.0  

[5 rows x 21 columns]

In [57]:
credit_data_merged.drop('STATUS', axis=1, inplace=True)


In [58]:
credit_data_merged

ID  CODE_GENDER  FLAG_OWN_CAR  FLAG_OWN_REALTY  CNT_CHILDREN  \
0       5008806            1             1                1             0   
1       5008806            1             1                1             0   
2       5008806            1             1                1             0   
3       5008806            1             1                1             0   
4       5008806            1             1                1             0   
...         ...          ...           ...              ...           ...   
537662  5150337            1             0                1             0   
537663  5150337            1             0                1             0   
537664  5150337            1             0                1             0   
537665  5150337            1             0                1             0   
537666  5150337            1             0                1             0   

        AMT_INCOME_TOTAL  NAME_INCOME_TYPE  NAME_EDUCATION_TYPE  \
0               112500.0                 4                    4   
1               112500.0                 4                    4   
2               112500.0                 4                    4   
3               112500.0                 4                    4   
4               112500.0                 4                    4   
...                  ...               ...                  ...   
537662          112500.0                 4                    4   
537663          112500.0                 4                    4   
537664          112500.0                 4                    4   
537665          112500.0                 4                    4   
537666          112500.0                 4                    4   

        NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  FLAG_MOBIL  FLAG_WORK_PHONE  \
0                        1                  1           1                0   
1                        1                  1           1                0   
2                        1                  1           1                0   
3                        1                  1           1                0   
4                        1                  1           1                0   
...                    ...                ...         ...              ...   
537662                   3                  4           1                0   
537663                   3                  4           1                0   
537664                   3                  4           1                0   
537665                   3                  4           1                0   
537666                   3                  4           1                0   

        FLAG_PHONE  FLAG_EMAIL  OCCUPATION_TYPE  CNT_FAM_MEMBERS   AGE  \
0                0           0               16              2.0  59.0   
1                0           0               16              2.0  59.0   
2                0           0               16              2.0  59.0   
3                0           0               16              2.0  59.0   
4                0           0               16              2.0  59.0   
...            ...         ...              ...              ...   ...   
537662           0           0                8              1.0  26.0   
537663           0           0                8              1.0  26.0   
537664           0           0                8              1.0  26.0   
537665           0           0                8              1.0  26.0   
537666           0           0                8              1.0  26.0   

        YEARS_EMPLOYED  MONTHS_BALANCE  APPROVAL_STATUS  
0                  4.0               0              1.0  
1                  4.0              -1              1.0  
2                  4.0              -2              1.0  
3                  4.0              -3              1.0  
4                  4.0              -4              1.0  
...                ...             ...              ...  
537662             4.0              -9             

In [59]:
 credit_approval_data = credit_data_merged.drop('ID', axis=1, inplace=False)

In [60]:
credit_approval_data

CODE_GENDER  FLAG_OWN_CAR  FLAG_OWN_REALTY  CNT_CHILDREN  \
0                 1             1                1             0   
1                 1             1                1             0   
2                 1             1                1             0   
3                 1             1                1             0   
4                 1             1                1             0   
...             ...           ...              ...           ...   
537662            1             0                1             0   
537663            1             0                1             0   
537664            1             0                1             0   
537665            1             0                1             0   
537666            1             0                1             0   

        AMT_INCOME_TOTAL  NAME_INCOME_TYPE  NAME_EDUCATION_TYPE  \
0               112500.0                 4                    4   
1               112500.0                 4                    4   
2               112500.0                 4                    4   
3               112500.0                 4                    4   
4               112500.0                 4                    4   
...                  ...               ...                  ...   
537662          112500.0                 4                    4   
537663          112500.0                 4                    4   
537664          112500.0                 4                    4   
537665          112500.0                 4                    4   
537666          112500.0                 4                    4   

        NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  FLAG_MOBIL  FLAG_WORK_PHONE  \
0                        1                  1           1                0   
1                        1                  1           1                0   
2                        1                  1           1                0   
3                        1                  1           1                0   
4                        1                  1           1                0   
...                    ...                ...         ...              ...   
537662                   3                  4           1                0   
537663                   3                  4           1                0   
537664                   3                  4           1                0   
537665                   3                  4           1                0   
537666                   3                  4           1                0   

        FLAG_PHONE  FLAG_EMAIL  OCCUPATION_TYPE  CNT_FAM_MEMBERS   AGE  \
0                0           0               16              2.0  59.0   
1                0           0               16              2.0  59.0   
2                0           0               16              2.0  59.0   
3                0           0               16              2.0  59.0   
4                0           0               16              2.0  59.0   
...            ...         ...              ...              ...   ...   
537662           0           0                8              1.0  26.0   
537663           0           0                8              1.0  26.0   
537664           0           0                8              1.0  26.0   
537665           0           0                8              1.0  26.0   
537666           0           0                8              1.0  26.0   

        YEARS_EMPLOYED  MONTHS_BALANCE  APPROVAL_STATUS  
0                  4.0               0              1.0  
1                  4.0              -1              1.0  
2                  4.0              -2              1.0  
3                  4.0              -3              1.0  
4                  4.0              -4              1.0  
...                ...             ...              ...  
537662             4.0              -9              1.0  
537663             4.0             -10              NaN  
537664             4.0             -11

# Training , Featuring Credit approval

In [61]:
from sklearn.model_selection import train_test_split

In [63]:
credit_approval_data.isnull().sum()


CODE_GENDER               0
FLAG_OWN_CAR              0
FLAG_OWN_REALTY           0
CNT_CHILDREN              0
AMT_INCOME_TOTAL          0
NAME_INCOME_TYPE          0
NAME_EDUCATION_TYPE       0
NAME_FAMILY_STATUS        0
NAME_HOUSING_TYPE         0
FLAG_MOBIL                0
FLAG_WORK_PHONE           0
FLAG_PHONE                0
FLAG_EMAIL                0
OCCUPATION_TYPE           0
CNT_FAM_MEMBERS           0
AGE                       0
YEARS_EMPLOYED            0
MONTHS_BALANCE            0
APPROVAL_STATUS        6965
dtype: int64

In [67]:
credit_approval_data = credit_approval_data.dropna()

In [68]:
credit_approval_data

CODE_GENDER  FLAG_OWN_CAR  FLAG_OWN_REALTY  CNT_CHILDREN  \
0                 1             1                1             0   
1                 1             1                1             0   
2                 1             1                1             0   
3                 1             1                1             0   
4                 1             1                1             0   
...             ...           ...              ...           ...   
537660            1             0                1             0   
537661            1             0                1             0   
537662            1             0                1             0   
537665            1             0                1             0   
537666            1             0                1             0   

        AMT_INCOME_TOTAL  NAME_INCOME_TYPE  NAME_EDUCATION_TYPE  \
0               112500.0                 4                    4   
1               112500.0                 4                    4   
2               112500.0                 4                    4   
3               112500.0                 4                    4   
4               112500.0                 4                    4   
...                  ...               ...                  ...   
537660          112500.0                 4                    4   
537661          112500.0                 4                    4   
537662          112500.0                 4                    4   
537665          112500.0                 4                    4   
537666          112500.0                 4                    4   

        NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  FLAG_MOBIL  FLAG_WORK_PHONE  \
0                        1                  1           1                0   
1                        1                  1           1                0   
2                        1                  1           1                0   
3                        1                  1           1                0   
4                        1                  1           1                0   
...                    ...                ...         ...              ...   
537660                   3                  4           1                0   
537661                   3                  4           1                0   
537662                   3                  4           1                0   
537665                   3                  4           1                0   
537666                   3                  4           1                0   

        FLAG_PHONE  FLAG_EMAIL  OCCUPATION_TYPE  CNT_FAM_MEMBERS   AGE  \
0                0           0               16              2.0  59.0   
1                0           0               16              2.0  59.0   
2                0           0               16              2.0  59.0   
3                0           0               16              2.0  59.0   
4                0           0               16              2.0  59.0   
...            ...         ...              ...              ...   ...   
537660           0           0                8              1.0  26.0   
537661           0           0                8              1.0  26.0   
537662           0           0                8              1.0  26.0   
537665           0           0                8              1.0  26.0   
537666           0           0                8              1.0  26.0   

        YEARS_EMPLOYED  MONTHS_BALANCE  APPROVAL_STATUS  
0                  4.0               0              1.0  
1                  4.0              -1              1.0  
2                  4.0              -2              1.0  
3                  4.0              -3              1.0  
4                  4.0              -4              1.0  
...                ...             ...              ...  
537660             4.0              -7              1.0  
537661             4.0              -8              1.0  
537662             4.0              -9

- Out of 1048575 records in credit_records, there are 530702 non-null entries that has matching ID in both application_records and credit_records

In [69]:
X = credit_approval_data.drop('APPROVAL_STATUS', axis=1)
y = credit_approval_data['APPROVAL_STATUS']
credit_approval_data['APPROVAL_STATUS'] = credit_approval_data['APPROVAL_STATUS'].astype('int')


/var/folders/ls/4_q0bc216xs15jl4fyc6skmw0000gp/T/ipykernel_88042/3751639179.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  credit_approval_data['APPROVAL_STATUS'] = credit_approval_data['APPROVAL_STATUS'].astype('int')


In [70]:
credit_approval_data

CODE_GENDER  FLAG_OWN_CAR  FLAG_OWN_REALTY  CNT_CHILDREN  \
0                 1             1                1             0   
1                 1             1                1             0   
2                 1             1                1             0   
3                 1             1                1             0   
4                 1             1                1             0   
...             ...           ...              ...           ...   
537660            1             0                1             0   
537661            1             0                1             0   
537662            1             0                1             0   
537665            1             0                1             0   
537666            1             0                1             0   

        AMT_INCOME_TOTAL  NAME_INCOME_TYPE  NAME_EDUCATION_TYPE  \
0               112500.0                 4                    4   
1               112500.0                 4                    4   
2               112500.0                 4                    4   
3               112500.0                 4                    4   
4               112500.0                 4                    4   
...                  ...               ...                  ...   
537660          112500.0                 4                    4   
537661          112500.0                 4                    4   
537662          112500.0                 4                    4   
537665          112500.0                 4                    4   
537666          112500.0                 4                    4   

        NAME_FAMILY_STATUS  NAME_HOUSING_TYPE  FLAG_MOBIL  FLAG_WORK_PHONE  \
0                        1                  1           1                0   
1                        1                  1           1                0   
2                        1                  1           1                0   
3                        1                  1           1                0   
4                        1                  1           1                0   
...                    ...                ...         ...              ...   
537660                   3                  4           1                0   
537661                   3                  4           1                0   
537662                   3                  4           1                0   
537665                   3                  4           1                0   
537666                   3                  4           1                0   

        FLAG_PHONE  FLAG_EMAIL  OCCUPATION_TYPE  CNT_FAM_MEMBERS   AGE  \
0                0           0               16              2.0  59.0   
1                0           0               16              2.0  59.0   
2                0           0               16              2.0  59.0   
3                0           0               16              2.0  59.0   
4                0           0               16              2.0  59.0   
...            ...         ...              ...              ...   ...   
537660           0           0                8              1.0  26.0   
537661           0           0                8              1.0  26.0   
537662           0           0                8              1.0  26.0   
537665           0           0                8              1.0  26.0   
537666           0           0                8              1.0  26.0   

        YEARS_EMPLOYED  MONTHS_BALANCE  APPROVAL_STATUS  
0                  4.0               0                1  
1                  4.0              -1                1  
2                  4.0              -2                1  
3                  4.0              -3                1  
4                  4.0              -4                1  
...                ...             ...              ...  
537660             4.0              -7                1  
537661             4.0              -8                1  
537662             4.0              -9

In [71]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)


In [72]:
from sklearn.tree import DecisionTreeClassifier


In [73]:
DT_model = DecisionTreeClassifier()


In [75]:
DT_model.fit(X_train, y_train)


DecisionTreeClassifier()

In [76]:
y_pred = DT_model.predict(X_test)


# Accuracy Score

In [78]:
from sklearn.metrics import accuracy_score

In [79]:
accuracy_score(y_test, y_pred)

0.9972049669934866

**0.9972049669934866**